src/aead/aead.rs

```diff
@@ -77,11 +77,14 @@ impl OpeningKey {
 
 /// Authenticates and decrypts (“opens”) data in place. When
 ///
-/// The input may have a prefix that is `in_prefix_len` bytes long; any such
+/// `nonce` must be unique for every use of the key to seal data; it must be
```

Comment:
(Don't let this PR be blocked over this discussion.)

> And if they choose the wrong nonce then there isn't any security issue because AES-GCM is designed to fail safe ...

That's a good point and it settles the discussion whether `open_in_place` needs to explain its nonces better. You're right, it's not a misuse resistance issue, only a minor usability problem.

About whether _ring_ should teach cryptography: not really. APIs should be usable in a safe way without background knowledge (because, if they're not, bad things happen unfortunately, see below). The best way coincides with your goal to make _ring_'s API impossible to misuse. In some cases, the API itself can't do it and explaining stuff in the documentation (or at least pointing to the right explanations, making people aware that problems exist) is plan B.
For example, the documentation should state that random nonces are not perfectly secure (and maybe point to sites that explain it). Without that pointer, users won't know about this. I didn't know and I've been working on crypto and crypto APIs for half a year now. Normal developers have half a day before their encryption has to work. So the documentation doesn't replace a book, but it replaces the need to read an entire book and all recent security developments before beginnig a project by pointing to the relevant stuff.

> they shouldn't be designing their own protocol
> They literally need to read a book.

I agree, they should. But some won't read a book and others won't understand it (because of lacking time or interest) and they still need to send encrypted data. They don't intend to design a protocol and they don't approach it like that, they just encrypt and go ahead.

There is a good example from the SSL area. This may be anecdotal evidence, but I think it's somewhat representative: dev turns off certificate validation, researcher informs about the vulnerability, dev says its secure because, on Wireshark, the traffic is obviously encrypted.  [http://android-ssl.org/files/p49.pdf http://dl.acm.org/citation.cfm?id=2516655 section 4.2.1]
That reasoning also applies to random nonces for AEAD: "Clever people say that AE is secure, I'm using secure randomness, it's visibly encrypted, done."

> _ring_ is designed to make it easy to verify that a ring-based implementation matches a specification of the higher-level protocol being implemented.

Fair enough.
Like all other crypto libraries, _ring_ can't avoid being used by uninformed users, too. And in fact, _ring_ is closest to being recommendable for such users, and getting better with every commit.